### PR TITLE
[ConvDiff] Add two flags to the solver

### PR DESCRIPTION
**Pull request by riccardotosi**
**Description**
This PR adds two flags to the solver: `use_orthogonal_subscales` and `time_integration_method`, and the capability of running the explicit elements merged in #7272.

Please mark the PR with appropriate tags: 
- Applications, Python

**Changelog**
Please summarize the changes in one list to generate the changelog:
- Adds the explicit scheme, to be able to run explicit elements recently merged
- Adds the two above-mentioned flags
- Tests added

As commented by @rubenzorrilla [#7272], we have moved the Python changes to a separate PR, since it modifies the parameters of the solver.

This PR adds the capability to run explicit schemes. The idea is to give priority to the explicit time integration with respect to previous methods (implicit, linear or nonlinear). This allowed to reduce to the minimum the number of changes of the solver. 

Any suggestion is welcome.

### CODE REVIEW

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (philbucher) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: [in StructuralMechanics this is called `time_integration_method`](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py#L12)

```suggestion
            "time_integration_method": "implicit",
```

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (philbucher) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: why doing this here?
I would say this is not the place

Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: Hi @philbucher , I copied from [the transient solver](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_transient_solver.py#L51).

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Ok.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Why are you adding this to the base solver if you are creating your own explicit one?

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Why are you adding this to the base solver if you are creating your own explicit one?

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Why are you adding this to the base solver if you are creating your own explicit one?

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_transient_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_transient_solver.py]: I think this was accidentally modified.

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]: This must be and elif statement. An error must be thrown in the else.

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]: In here:
1. Remove "implicit / not defined". It is either implicit or not defined, but no both at the same time.
2. If the `Has` is `true` save the provided `time_integration_method`. If not, set the `implicit` as default and
  2.1 Throw a warning specifying this
  2.2 And add it to the solver_settings dict (not only the auxiliary variable)

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]: This must be an `elif` statement. An `else` with an error saying that the provided `time_integration_method` is wrong must be placed below.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: If you don't switch off the OSS this must be an error.

Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]:
```suggestion
                KratosMultiphysics.Logger.PrintWarning(self.__class__.__name__, "The selected element ", element_name, "does not support OSS projection. Select QSConvectionDiffusionExplicit or DConvectionDiffusionExplicit instead.")
```

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: This does nothing.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: This is something specific of the explicit element. I think that at this point it must be moved to the explicit new solver.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Same comment. The unique elements that support this are the explicit ones. I think it should be placed in the explicit defaults.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py:523]: We should find a way to improve this. Note that each new element pollutes the base solver. Can the same behaviour be achieved by introducing an auxiliary function?

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
Some concerns that affect the structure of the current solvers.

Also the changes should be merged in accordance to the GUI ones.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Moving to explicit solver.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Moving to explicit solver.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Moving to explicit solver.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_transient_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_transient_solver.py]: Yes, should be reverted now.

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]: Ok, even though the error can never happen since by default we set `implicit`.

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]: Ok, thanks. I was unsure about the best strategy to follow.

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]: Ok, even though setting `implicit` by default will never happen. Done.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: I will not apply this suggestion since now I throw an error, as suggested by you.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Added here an error.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Ok, now this is an error.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Ok. I will add the flag "use_orthogonal_subscales" to the explicit solver. That should not be difficult to move in the base_solver when there will be other OSS-compatible elements.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py]: Ok.

**Comment by riccardotosi** [comment]
> 
> 
> Some concerns that affect the structure of the current solvers.
> 
> Also the changes should be merged in accordance to the GUI ones.

Hi @rubenzorrilla, I applied your suggestions. Please also note that I had to add `time_integration_method` to the default settings of `python_scripts/adjoint_diffusion_solver.py`. Additionally, this solver does is not derived from `ConvectionDiffusionBaseSolver`.

Let me know if there are other comments.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py:523]: What comes to my mind is to define `_get_element_condition_replace_settings` inside each solver. This way, we do not pollute the base solver. What do you think?

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (philbucher) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: hm why not doing this in `Initialize`?

Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: I have no problem, but I would like to listen to @KratosMultiphysics/fluid-maintainers @rubenzorrilla opinions before doing this, since I am not introducong a new feature, but just being consistent with previous codes.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py
Comment (rubenzorrilla) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py:523]: Something of the sort. What we do in the `fluid_solver` is to define a `_ReplaceElementsAndConditions` method.

Comment (rubenzorrilla) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_base_solver.py:523]: In any case we can do it in a future PR as this would imply changing all the solvers.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]:
```suggestion
        super().__init__(model, custom_settings)
```

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]:
```suggestion
from KratosMultiphysics.ConvectionDiffusionApplication import convection_diffusion_base_solver
```

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]:
```suggestion
class ConvectionDiffusionExplicitSolver(convection_diffusion_base_solver.ConvectionDiffusionBaseSolver):
```

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]:
```suggestion
        element_name = self.settings["element_replace_settings"]["element_name"].GetString()
```

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: I think this is not required. It will appear when the error pop up.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]:
```suggestion
            oss_element_list = [QSConvectionDiffusionExplicit", "DConvectionDiffusionExplicit"]
            if element_name in oss_element_list:
```

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: If I'm not wrong this block is not required (`OSS_SWITCH` is zero by default)'. Indeed, you are throwing an error saying that the element does not support OSS projections when the user selects OSS projections false.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]:
```suggestion
        KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Construction finished")
```
Please, increase your maximum characters per line. I think in this case it makes readability much more difficult.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: I am using a formatter which automatically splits lines if too long. Is there a default Python formatter? I have not found any in the Style guide.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: Why an extra level for a single field? Furthermore, I think that `use_orthogonal_subscales` and `dynamic_tau` must be at the same level.

If you want to encapsulate all the formulation-related settings we can create an extra level called `formulation` as we did in the fluid solvers.However this needs to be done in a separate PR as it would affect the entire application and the GUI.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: Is there a default?

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py:73]:
```suggestion
    def _CreateBuilderAndSolver(self):
```
Style guide. Apply same suggestion to the other methods.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]:
```suggestion
        convection_diffusion_scheme = self.get_solution_scheme() # to call _create_solution_scheme method of the explicit solver
```

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: You are not creating the solution scheme in here. Indeed the explicit strategies have no scheme as the strategy is the scheme itself. Please remove this. Indeed I think that we must throw an error in here.

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]:
```suggestion
        KratosMultiphysics.Logger.PrintWarning("Time integration method was not provided. Setting \'implicit\' as default.")
```

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]:
```suggestion
    if not solver_settings.Has("time_integration_method"):
```

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]:
```suggestion
    time_integration_method = solver_settings["time_integration_method"].GetString()
```

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py]: You can use `.format` to make this implementation cleaner.

File: applications/ConvectionDiffusionApplication/tests/test_convection_diffusion_explicit_element.py
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/tests/test_convection_diffusion_explicit_element.py]: I came up with the idea of calling this `TestConvectionDiffusionBar` so we can reuse this same test also for the non-explicit elements. What do you think? 

It should be as easy as adding more `testMyElement` functions together with the corresponding `ProjectParameters.json` files. We can add the implicit element tests in a future PR.

**Review by rubenzorrilla — COMMENTED** [review]
Overall it looks great now. Just a few minor comments.

Approving once the changes in the GUI have been done.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: Ok, removed.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: I did remove the error message.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: People says 80. But I think this is something that comes from ancient screens. I'd say "common sense" to ensure that going through the code is as comfortable as possible.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: Removed extra level.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py:73]: The convection diffusion base solver uses `_create_builder_and_solver`... So I cannot apply these changes.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py]: I removed the method.

File: applications/ConvectionDiffusionApplication/tests/test_convection_diffusion_explicit_element.py
Comment (riccardotosi) [reply applications/ConvectionDiffusionApplication/tests/test_convection_diffusion_explicit_element.py]: Ok, done.

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py
Comment (rubenzorrilla) [reply applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_explicit_solver.py:73]: We should change it at some point.

**Review by rubenzorrilla — APPROVED** [review]
I'll add the new formulations to the GUI soon.

**Pull request merged**
